In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
from scipy import stats 

1. Buscar un conjunto de datos a analizar:
Se recomienda que el conjunto de datos a analizar tenga variables numéricas y categóricas. Ya que también se analizarán.

In [4]:
# Hemos elegido un conjunto de datos que recoge el salario de cada posicion en el sector de data science, por año, por país y otros.

In [5]:
df = pd.read_csv('ds_salaries.csv', index_col = 0)
df.head(3)

work_year experience_level employment_type                   job_title  \
0       2020               MI              FT              Data Scientist   
1       2020               SE              FT  Machine Learning Scientist   
2       2020               SE              FT           Big Data Engineer   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   70000             EUR          79833                 DE             0   
1  260000             USD         260000                 JP             0   
2   85000             GBP         109024                 GB            50   

  company_location company_size  
0               DE            L  
1               JP            S  
2               GB            M

In [6]:
df.info()
# Vemos que tenemos solo 4 columnas de tipo numérico

<class 'pandas.core.frame.DataFrame'>
Int64Index: 607 entries, 0 to 606
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           607 non-null    int64 
 1   experience_level    607 non-null    object
 2   employment_type     607 non-null    object
 3   job_title           607 non-null    object
 4   salary              607 non-null    int64 
 5   salary_currency     607 non-null    object
 6   salary_in_usd       607 non-null    int64 
 7   employee_residence  607 non-null    object
 8   remote_ratio        607 non-null    int64 
 9   company_location    607 non-null    object
 10  company_size        607 non-null    object
dtypes: int64(4), object(7)
memory usage: 56.9+ KB


>
> `Enlace a la leyenda del Dataset:`
>
>  https://www.kaggle.com/datasets/ruchi798/data-science-job-salaries?resource=download
>

2. Extraer los siguentes estadísiticos del conjunto de datos para varias de las variables disponibles:
Media,
Desviación con respecto de la media, 
Varianza,
Desviación típica,
Moda,
Mediana,
Media ponderada,
Robustez,
Coeficiente de variación

In [7]:
# Calculamos la media de dos columnas, salary in usd (para que todos los salarios nos den en la misma moneda) y work year

media_salary= df['salary_in_usd'].mean()
media_salary

112297.86985172982

In [8]:
media_year = df['work_year'].mean()
media_year

2021.4052718286655

In [36]:
# Calculamos la desviacion con respecto a la media de las mismas columnas

salary= df['salary_in_usd']
desviacion_media= np.abs(salary - media_salary )
desviacion_media

0       32464.869852
1      147702.130148
2        3273.869852
3       92297.869852
4       37702.130148
           ...      
602     41702.130148
603     13702.130148
604     16702.130148
605     37702.130148
606     87702.130148
Name: salary_in_usd, Length: 607, dtype: float64

In [37]:
work_year = df['work_year']
desviacion_media_year = np.abs(work_year - media_year)
desviacion_media_year

0      1.405272
1      1.405272
2      1.405272
3      1.405272
4      1.405272
         ...   
602    0.594728
603    0.594728
604    0.594728
605    0.594728
606    0.594728
Name: work_year, Length: 607, dtype: float64

In [38]:
# Calculamos la varianza

var_salary = np.var(salary)
var_salary

5026637881.193936

In [39]:
var_year = np.var(work_year)
var_year

0.47825886350621105

In [40]:
# Calculamos la desviación típica

std_salary = np.std(salary)
std_salary

70898.78617574448

In [41]:
std_year = np.std(work_year)
std_year

0.6915626244283384

In [82]:
# Calculamos la moda

moda1, frecuencia1 = stats.mode(salary)
print('la moda es :', moda1)
print('la frecuencia es :', frecuencia1)

la moda es : [100000]
la frecuencia es : [15]


In [83]:
moda2, frecuencia2 = stats.mode(work_year)
print('la moda es :', moda2)
print('la frecuencia es :', frecuencia2)

la moda es : [2022]
la frecuencia es : [318]


In [44]:
# Calculamos la mediana

print('la mediana del salario es: ', np.median(salary))
print('la mediana del año es: ', np.median(work_year))

la mediana del salario es:  101570.0
la mediana del año es:  2022.0


In [45]:
# Calculamos la media ponderada sin pesos. 
# En este caso el resultado es igual que la media.

print('la media ponderada del salario es: ', np.average(salary))
print('la media ponderada del año es: ', np.average(work_year))

la media ponderada del salario es:  112297.86985172982
la media ponderada del año es:  2021.4052718286655


In [58]:
# Imaginamos que queremos darle más peso a los registros realizados en el año 2022
## Nos creamos un array con los años y otro con los pesos.

arr_years = np.array(work_year)
weights = np.where(arr_years == 2022, 0.7, 0.15)

In [60]:
print('la media ponderada del año es: ', np.average(arr_years, weights= weights))

la media ponderada del año es:  2021.7963902989277


In [77]:
# Robustez

# Para comprobar la robustez de los cálculos estadísticos tenemos que probar a modificar los datos de nuestras variables
# y ver cuánto se ven afectados los resultados estadísticos. Cuanto menos les afecten los cambios, más robustos son.

# Hacemos una copia para modificar algunos de los datos
dfsalary = df['salary_in_usd'].copy()
dfsalary.head()

0     79833
1    260000
2    109024
3     20000
4    150000
Name: salary_in_usd, dtype: int64

In [78]:
# Eliminamos todos los valores por debajo de 60.000 $
dfsalary.where(dfsalary > 6000, 0, inplace = True)

In [84]:
# Calculamos la media, la mediana y la moda
print('la media de dfsalary es: ', np.mean(dfsalary))

moda3, frecuencia3 = stats.mode(dfsalary)
print('la moda es :', moda3)
print('la frecuencia es :', frecuencia3)

print('la mediana del dfsalary es: ', np.median(dfsalary))


la media de dfsalary es:  112242.62108731466
la moda es : [100000]
la frecuencia es : [15]
la mediana del dfsalary es:  101570.0


In [86]:
# Comparamos los cambios

print('la media del salario antes de modificar era:', np.mean(salary))
print('la media del salario después de modificar es:', np.mean(dfsalary))
print('Ha cambiado', abs(np.mean(salary) - np.mean(dfsalary)), ' unidades con las modificaciones')

print('-------------------------------------')

print('la mediana del salario antes de modificar era:', np.median(salary))
print('la mediana del salario después de modificar es:', np.median(dfsalary))
print('Ha cambiado', abs(np.median(salary)- np.median(dfsalary)), ' unidades con las modificaciones')

print('-------------------------------------')

print('la moda y frecuencia del salario antes de modificar era:', moda1, ',', frecuencia1)
print('la moda y frecuencia del salario después de modificar es:', moda3, ',', frecuencia3)
print('Ha cambiado', abs(moda1 - moda3), ' con las modificaciones')

la media del salario antes de modificar era: 112297.86985172982
la media del salario después de modificar es: 112242.62108731466
Ha cambiado 55.248764415155165  unidades con las modificaciones
-------------------------------------
la mediana del salario antes de modificar era: 101570.0
la mediana del salario después de modificar es: 101570.0
Ha cambiado 0.0  unidades con las modificaciones
-------------------------------------
la moda y frecuencia del salario antes de modificar era: [100000] , [15]
la moda y frecuencia del salario después de modificar es: [100000] , [15]
Ha cambiado [0]  con las modificaciones


In [ ]:
# CONCLUSIONES: La moda y la mediana son cálculos estadísticos mucho más robustos que la media, 
# ya que la media utuliza todos los valores para calcular el resultado.

In [61]:
# Calculamos el coeficiente de variación

print('el coeficiente de variación del salario es: ', std_salary / media_salary)
print('el coeficiente de variación del año es: ', std_year / media_year)


el coeficiente de variación del salario es:  0.63134577948232
el coeficiente de variación del año es:  0.00034211972931223034


3. Interpretación de los resultados.
Ahora interpreta los resultados obtenidos de los diferentes estadísticos estudiados haciendo hincapié en aspectos interesantes de tus datos y explica lo mejor posible la naturaleza de los mismos.


##### `salary_in_usd`


En el caso de los valores de los salarios, vemos que la **varianza** tiene un valor muy elevado y, aunque la **diferencia entre la media y la mediana** es de `10.727,86$`, deducimos que tenemos un _conjunto de valores bastante disperso_. 

> Esto apunta a que las condiciones de salario varían mucho dependiendo de factores como el país, la experiencia, el tipo de contrato, etc.

Además, de acuerdo a la **moda**, vemos que el salario más repetido entre los encuestados es de `100.000$`, con una **frecuencia** de `15` entre `607` registros, y se acerca bastante a la media (media_salary = 112297.87 $).

##### `work_year`


En cuanto al tiempo en años, vemos que hay _poca variación_, eso significa que los datos recogidos están distribuidos de forma bastante uniforme en el tiempo, que oscula en un rango de 3 años. 

Por eso la media y la mediana se diferencian en un `0.6`, osea `6` meses.

> Sin embargo, de acuerdo a la **moda**, el año con más registros es el `2022`, con una **frecuencia** de `318` sobre `607`, es decir un _poco más de la mitad de los registros corresponden al año 2022_. 